File: amazon/amazon_avg.scala
import org.apache.spark.SparkContext
import org.apache.spark.SparkContext._
import org.apache.spark.SparkConf

object AmazonAVG {
  def main(args: Array[String]) {

    val sc = new SparkContext(new SparkConf().setAppName("Spark Count")))

    if(args.length < 1){
      println("No input file provided")
      System.exit(1)
    }

    val lines = sc.textFile(args(0))

    // do some standard mapping
    val rows = lines.map(a => a.split("\016"))
    val mappi = rows.map(x => (x(0),x))

    //var p2 = mappi.combineByKey(x => 0.0, (x:Double,v:Array[String]) => x, (v:Double,v2:Double) => v + v2

      // reduce to (rating, ratingsPerItem) -> calc avg
      var ratings = mappi.combineByKey(x => (x(6).toDouble, 1), (z:(Double,Int),v) => (z._1 + v(6).toDouble, z._2 + 1), (v:(Double, Int),v2:(Double,Int)) => (v._1 + v2._1, v._2 + v2._2)).mapValues(x => x._1 / x._2)

      // resolve tuple to avg
      val avgRating = ratings.mapValues("%1.2f".format(_))
      val totalMean = ratings.map(_._2).mean()

      println(totalMean)

    }

  }

File: amazon/amazon_avg.py
#!/usr/bin/python2

import sys
from pyspark import SparkContext

sc = SparkContext()

if len(sys.argv) < 2:
    print("no input file specified")
    sys.exit(1)

inputFile = sys.argv[1]

print("input %s" % inputFile)

lines = sc.textFile(inputFile)
rows = lines.map(lambda x: x.split("\016"))
mappi = rows.map(lambda x: (x[0], x))

ratings = mappi.combineByKey(lambda x: (float(x[6]),1), lambda x,y: (x[0] + float(y[6]), x[1] + 1), lambda x,y: (x[0] + y[0], x[1] + y[1] )).mapValues(lambda x: x[0] / x[1])

avgRating = ratings.mapValues(lambda x: "{:1.2f}".format(x))
totalMean = ratings.map(lambda x: x[1]).mean()
print(totalMean)
